File: src/main/java/supercritical/api/nuclear/fission/FissionFuelRegistry.java
package supercritical.api.nuclear.fission;

import java.util.Collection;
import java.util.Map;

import net.minecraft.item.ItemStack;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import gregtech.api.util.ItemStackHashStrategy;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenCustomHashMap;
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;

public class FissionFuelRegistry {

    private static final Map<String, IFissionFuelStats> IDENTIFIED_FUELS = new Object2ObjectOpenHashMap<>();
    private static final Map<ItemStack, IFissionFuelStats> FUELS = new Object2ObjectOpenCustomHashMap<>(
            ItemStackHashStrategy.comparingAllButCount());

    public static void registerFuel(@NotNull ItemStack item, @NotNull IFissionFuelStats fuel) {
        IDENTIFIED_FUELS.put(fuel.getId(), fuel);
        FUELS.put(item, fuel);
    }

    @Nullable
    public static IFissionFuelStats getFissionFuel(ItemStack stack) {
        return FUELS.get(stack);
    }

    @NotNull
    public static Collection<ItemStack> getAllFissionableRods() {
        return FUELS.keySet();
    }

    @Nullable
    public static IFissionFuelStats getFissionFuel(String name) {
        return IDENTIFIED_FUELS.get(name);
    }

}

File: src/main/kotlin/supercritical/api/nuclear/fission/FissionFuelRegistry.kt
package supercritical.api.nuclear.fission

import com.gregtechceu.gtceu.utils.ItemStackHashStrategy
import it.unimi.dsi.fastutil.objects.Object2ObjectOpenCustomHashMap
import net.minecraft.world.item.ItemStack
import java.util.*

object FissionFuelRegistry {
    private val IDENTIFIED_FUELS: MutableMap<String?, IFissionFuelStats?> = linkedMapOf()
    private val FUELS: MutableMap<ItemStack, IFissionFuelStats?> = Object2ObjectOpenCustomHashMap(ItemStackHashStrategy.comparingAllButCount())

    fun registerFuel(item: ItemStack, fuel: IFissionFuelStats) {
        IDENTIFIED_FUELS[fuel.id] = fuel
        if (!item.isEmpty) {
            FUELS[item.copyWithCount(1)] = fuel
        }
    }

    fun registerFuel(fuel: IFissionFuelStats) {
        IDENTIFIED_FUELS[fuel.id] = fuel
    }

    fun getFissionFuel(stack: ItemStack): IFissionFuelStats? {
        if (stack.isEmpty) return null
        return FUELS[stack]
    }

    val allFissionableRods: MutableCollection<ItemStack>
        get() = Collections.unmodifiableSet<ItemStack>(FUELS.keys)

    fun getFissionFuel(name: String?): IFissionFuelStats? = IDENTIFIED_FUELS[name]

    val allFuelStats: MutableCollection<IFissionFuelStats?>
        get() = Collections.unmodifiableCollection<IFissionFuelStats?>(IDENTIFIED_FUELS.values)
}
